packages/agent/src/upgrader-utils.ts

```diff
@@ -0,0 +1,161 @@
+import { normalizeErrorString } from '@medplum/core';
+import { createWriteStream } from 'node:fs';
+import { platform } from 'node:os';
+import { resolve } from 'node:path';
+import { Readable } from 'node:stream';
+
+export type ReleaseManifest = { tag_name: string; assets: { name: string; browser_download_url: string }[] };
+export type SupportedOs = 'windows' | 'linux';
+
+export const UPGRADE_MANIFEST_PATH = resolve(__dirname, 'upgrade.json');
+export const UPGRADER_LOG_PATH = resolve(
+  __dirname,
+  `upgrader-logs-${new Date().toISOString().replace(/:\s*/g, '-')}.txt`
+);
+export const GITHUB_RELEASES_URL = 'https://api.github.com/repos/medplum/medplum/releases';
+export const RELEASES_PATH = resolve(__dirname);
+
+const releaseManifests = new Map<string, ReleaseManifest>();
+
+export function clearReleaseCache(): void {
+  releaseManifests.clear();
+}
+
+export function getOsString(): SupportedOs {
+  switch (platform()) {
+    case 'win32':
+      return 'windows';
+    case 'linux':
+      return 'linux';
+    default:
+      throw new Error(`Unsupported platform: ${platform()}`);
+  }
+}
+
+export function isValidSemver(version: string): boolean {
+  return /^\d+\.\d+\.\d+$/.test(version);
+}
+
+export function assertReleaseManifest(candidate: unknown): asserts candidate is ReleaseManifest {
+  const manifest = candidate as ReleaseManifest;
+  if (!manifest.tag_name) {
+    throw new Error('Manifest missing tag_name');
+  }
+  const assets = manifest.assets;
+  if (!assets?.length) {
+    throw new Error('Manifest missing assets list');
+  }
+  for (const asset of assets) {
+    if (!asset.browser_download_url) {
+      throw new Error('Asset missing browser download URL');
+    }
+    if (!asset.name) {
+      throw new Error('Asset missing name');
+    }
+  }
+}
+
+export async function checkIfValidMedplumVersion(version: string): Promise<boolean> {
+  if (!isValidSemver(version)) {
+    return false;
+  }
+  try {
+    await fetchVersionManifest(version);
+  } catch (_err) {
+    return false;
+  }
+  return true;
+}
+
+export async function fetchLatestVersionString(): Promise<string> {
+  const latest = await fetchVersionManifest();
+  if (!latest.tag_name.startsWith('v')) {
+    throw new Error(`Invalid release name found. Release tag '${latest.tag_name}' did not start with 'v'`);
+  }
+  return latest.tag_name.slice(1);
+}
+
+export async function downloadRelease(version: string, path: string): Promise<void> {
+  const release = await fetchVersionManifest(version);
+
+  // Get download url
+  const downloadUrl = parseDownloadUrl(release, getOsString());
+
+  // Write file to RELEASE_INSTALLER_FOLDER
+  const { body } = await fetch(downloadUrl);
+  if (!body) {
+    throw new Error('Body not present on Response');
+  }
+
+  // @ts-expect-error Slight type mismatch
```

Comment:
Can we use cast rather than `@ts-expect-error`?

Reply:
Originally tried assert, didn't work naively. But I found that asserting `ReadableStream` from `node:stream/web` worked 👍 done